File: app/src/main/java/com/geridea/trentastico/utils/time/WeekTime.java
package com.geridea.trentastico.utils.time;


/*
 * Created with ♥ by Slava on 15/03/2017.
 */

import java.util.Calendar;

public class WeekTime {
    private int year;
    private int weekNumber;

    public WeekTime(int year, int weekNumber) {
        this.year = year;
        this.weekNumber = weekNumber;
    }

    public WeekTime() {
        initFromCalendar(Calendar.getInstance());
    }

    public WeekTime(Calendar calendar) {
        initFromCalendar(calendar);
    }

    public WeekTime(long milliseconds) {
        Calendar dummy = CalendarUtils.getClearDummyCalendar();
        dummy.setTimeInMillis(milliseconds);

        initFromCalendar(dummy);
    }

    public WeekTime(WeekDayTime weekDayTime) {
        this(weekDayTime.getYear(), weekDayTime.getWeekNumber());
    }

    public void initFromCalendar(Calendar calendar) {
        year       = calendar.get(Calendar.YEAR);
        weekNumber = calendar.get(Calendar.WEEK_OF_YEAR);
    }

    public int getYear() {
        return year;
    }

    public int getWeekNumber() {
        return weekNumber;
    }

    /**
     * @return true if the current week time start after the argument.
     */
    public boolean after(WeekTime arg) {
        if (this.year == arg.year) {
            return this.weekNumber > arg.weekNumber;
        } else
            return this.year > arg.year;
    }

    /**
     * @return true if the current week time start before the argument.
     */
    public boolean before(WeekTime arg) {
        if (this.year == arg.year) {
            return this.weekNumber < arg.weekNumber;
        } else
            return this.year < arg.year;
    }


    /**
     * Creates a new <code>WeekTime</code> that is referring to the current week.
     */
    public static WeekTime getCurrentInstance() {
        return new WeekTime(CalendarUtils.getCachedToday());
    }

    public boolean hasSameWeekTime(WeekDayTime day) {
        return getWeekNumber() == day.getWeekNumber() &&
                     getYear() == day.getYear();
    }

    public boolean before(WeekDayTime day) {
        if (this.year == day.getYear()) {
            return this.weekNumber < day.getWeekNumber();
        } else
            return this.year < day.getYear();
    }

    public boolean after(WeekDayTime day) {
        if (this.year == day.getYear()) {
            return this.weekNumber > day.getWeekNumber();
        } else
            return this.year > day.getYear();
    }


    public void addWeeks(int numWeeksToAdd) {
        //Some years may have 53 weeks, so it's a good thing to delegate this calculation to the
        //calendar
        Calendar dummy = CalendarUtils.getClearDummyCalendar();
        dummy.set(Calendar.YEAR, year);
        dummy.set(Calendar.WEEK_OF_YEAR, weekNumber);
        dummy.add(Calendar.WEEK_OF_YEAR, numWeeksToAdd);

        initFromCalendar(dummy);
    }

    public WeekTime copy() {
        return new WeekTime(year, weekNumber);
    }

    @Override
    public boolean equals(Object obj) {
        if (obj instanceof WeekTime) {
            WeekTime time = (WeekTime) obj;
            return time.weekNumber == weekNumber && time.year == year;
        }
        return false;
    }

    /**
     * True if the current week time start before or equals the argument.
     */
    public boolean beforeOrEqual(WeekTime weekTime) {
        return before(weekTime) || equals(weekTime);
    }

    /**
     * True if the current week time start after or equals the argument.
     */
    public boolean afterOrEqual(WeekTime weekTime) {
        return after(weekTime) || equals(weekTime);
    }

    @Override
    public String toString() {
        return year+"/"+weekNumber;
    }
}

File: app/src/main/java/com/geridea/trentastico/utils/time/WeekTime.kt
package com.geridea.trentastico.utils.time


/*
 * Created with ♥ by Slava on 15/03/2017.
 */

import java.util.*

class WeekTime {
    private var year: Int = 0
        private set
    private var weekNumber: Int = 0
        private set

    constructor(year: Int, weekNumber: Int) {
        this.year = year
        this.weekNumber = weekNumber
    }

    constructor(calendar: Calendar) {
        year       = calendar.get(Calendar.YEAR)
        weekNumber = calendar.get(Calendar.WEEK_OF_YEAR)
    }

    /**
     * @return true if the current week time start after the argument.
     */
    fun after(arg: WeekTime): Boolean = if (this.year == arg.year) {
        this.weekNumber > arg.weekNumber
    } else
        this.year > arg.year

    /**
     * @return true if the current week time start before the argument.
     */
    fun before(arg: WeekTime): Boolean = if (this.year == arg.year) {
        this.weekNumber < arg.weekNumber
    } else
        this.year < arg.year

    fun hasSameWeekTime(day: WeekDayTime): Boolean = weekNumber == day.weekNumber && year == day.year

    fun before(day: WeekDayTime): Boolean = if (this.year == day.year) {
        this.weekNumber < day.weekNumber
    } else
        this.year < day.year

    fun after(day: WeekDayTime): Boolean = if (this.year == day.year) {
        this.weekNumber > day.weekNumber
    } else
        this.year > day.year


    fun addWeeks(numWeeksToAdd: Int): WeekTime {
        //Some years may have 53 weeks, so it's a good thing to delegate this calculation to the
        //calendar
        val dummy = CalendarUtils.clearCalendar
        dummy.set(Calendar.YEAR, year)
        dummy.set(Calendar.WEEK_OF_YEAR, weekNumber)
        dummy.add(Calendar.WEEK_OF_YEAR, numWeeksToAdd)

        year = dummy.get(Calendar.YEAR)
        weekNumber = dummy.get(Calendar.WEEK_OF_YEAR)

        return this
    }

    fun copy(): WeekTime = WeekTime(year, weekNumber)

    override fun equals(other: Any?): Boolean {
        if (other is WeekTime) {
            val time = other as WeekTime?
            return time!!.weekNumber == weekNumber && time.year == year
        }
        return false
    }

    /**
     * True if the current week time start before or equals the argument.
     */
    fun beforeOrEqual(weekTime: WeekTime): Boolean = before(weekTime) || equals(weekTime)

    override fun toString(): String = year.toString() + "/" + weekNumber

    override fun hashCode(): Int {
        var result = year
        result = 31 * result + weekNumber
        return result
    }

    val millis: Long
        get() {
            val calendar = CalendarUtils.clearCalendar
            calendar.set(Calendar.WEEK_OF_YEAR, weekNumber)
            calendar.set(Calendar.YEAR, year)

            return calendar.timeInMillis
        }
}
